### DISCUSSION TITLE
How can I get more control by using SPUStandardUpdaterController?

### DISCUSSION THREAD

**Discussion by AreroKetahi** [sparkle-project/Sparkle#2461] [Q&A]
Hello:

I am using Sparkle 2.5.1 to update my app, which is a SwiftUI app, provided by SwiftUI Lifecycle.

I use my custom update view, rather than the spark of the default view, and keep the sparks check the update and download the update background (using `checkForUpdatesInBackground ()`), to ensure that the update messages can be sent to my custom view is correct.

Now, I want my app to display the download progress, but I can't find any way to get a value for the download progress.

After extraction, I wanted my customizations to call the install update feature immediately, instead of calling the Sparkle Update dialog box and pressing it again, which I thought was too fragmented an experience.

How can I implement these features in my app?

**Comment by zorgiepoo** [comment]
If you want a custom user interface instead of Sparkle's standard user interface for showing the update to the user you will need your own [SPUUserDriver](https://sparkle-project.org/documentation/api-reference/Protocols/SPUUserDriver.html) and need to instantiate a [SPUUpdater](https://sparkle-project.org/documentation/api-reference/Classes/SPUUpdater.html) with it. (Sounds like what you're asking).

If you want to download the update automatically in the background without disrupting the user you should use [SUAutomaticallyUpdate](https://sparkle-project.org/documentation/customization/).

As discussed in #2452 do note before extraction in some cases the user may be asked to authorize installing the update (if needed, depending on how their system is set up).

**Comment by AreroKetahi** [comment]
Now I can have my app download updates in the background! But I still can't get the download progress in real time. 

I used a `ProgressView` to show the download progress, but I think such an uncertain download will make the user uneasy. 

How can I get the download progress in real time and how to start the update directly through code.

**Comment by zorgiepoo** [comment]
You can show download progress in your custom view if you have your own `SPUUserDriver`. But that does not show if automatic updates (`SUAutomaticallyUpdate`) are enabled so maybe you don't want that.

**Comment by AreroKetahi** [comment]
I understand `SPUUserDriver` can custom what will show to user. But I don't know how to do it in SwiftUI, because AppKit and SwiftUI have less same. Could you please give me some example to explain it? I feel nervous about it. Thanks!

**Comment by zorgiepoo** [comment]
I am not that well-versed with SwiftUI. [Here is an untested template](https://gist.github.com/zorgiepoo/7508a37f08ffb4b9fc0cb92347e1845c). You will need to fill in the gaps. For any of the Sparkle related APIs, refer to their documentation.

**Comment by AreroKetahi** [comment]
Thanks your example, I understand most of it, but I do have a question, how can I handling error in my custom user driver. Now, once error caused, `checkForUpdates()` and `checkForUpdatesInBackground()` won't work when I call them again. How can I solve this problem.

**Reply by zorgiepoo** [reply]
Did you implement  showUpdaterError in my template and acknowledge the error?

**Reply by AreroKetahi** [reply]
I implement it by passing the error to a '@Observable' class (because of SwiftUI) and a view will appear.

Here's a demo I implemented:
```swift
func showUpdaterError(_ error: error, acknowledgement: @escaping () -> Void) {
    controller.status =.error
    controller.error = error
    controller.isErrorPresent = true
    acknowledgement()
}
```

I'm new to Objective-C, so I don't fully understand how `acknowledgement()` works. I apologize for that.

**Reply by zorgiepoo** [reply]
it's just a completion handler to tell Sparkle when you're done presenting the error (I should rename acknowledgement to completionHandler or reply ..)

**Reply by zorgiepoo** [reply]
The API documentation for checkForUpdates and checkForUpdatesInBackground explain when they can/can't be called.
